opentracing/dapperish/main.go

```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"html"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"runtime"
+
+	"github.com/opentracing/api-golang/opentracing"
+)
+
+func client() {
+	reader := bufio.NewReader(os.Stdin)
+	for {
+		span, ctx := opentracing.StartSpan("getInput")
+		span.Info("ctx: ", ctx)
+		fmt.Print("Enter text: ")
+		text, _ := reader.ReadString('\n')
+		span.Info(text)
+
+		httpClient := &http.Client{}
+		httpReq, _ := http.NewRequest("POST", "http://localhost:8080/", bytes.NewReader([]byte(text)))
+		opentracing.AddContextIDToHttpHeader(span.ContextID(), httpReq.Header)
+		resp, err := httpClient.Do(httpReq)
+		if err != nil {
+			span.Error("error: ", err)
+		} else {
+			span.Info("got response: ", resp)
+		}
+
+		span.Finish()
+	}
+}
+
+func server() {
+	http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
+		reqCtxID, err := opentracing.GetContextIDFromHttpHeader(
+			req.Header, opentracing.GlobalTracer())
+		if err != nil {
+			panic(err)
+		}
+
+		serverSpan, _ := opentracing.StartSpan("serverSpan", reqCtxID)
```

Comment:
> obviously each server span is a blocking child of the corresponding client span

That's actually different from what I was thinking. A naive way to model this would be with fully nested spans
```
* start client span
** start router1 server span
*** start router1 client span (as client of router2)
**** start router2 server span
***** start router2 client span (as client of target service)
****** start actual service server span
```
This is perfectly doable, but looses the dependency between top and bottom services. I consider hops through the routers to be a part of the overall RPC call, thus they are not nested. And unlike  regular service spans that start on receipt of request and end on emitting a response, the router spans start on receipt of request and end on sending another request to the next hop, i.e. they represent the time spent in the router while passing on the message. To reuse my previous diagram:
```
 client span
+----------------------------------------------------------------------------+
    +------+   +------+   +-------------------+ +----------+ +------------+
     hop 1      hop 2      server                back-hop-2   back-hop-1
```

>> Span.AddCausalityHint(opentracing.FutureChildrenDependOnPreviouslyCompletedChildren)

But how would parent know the ordering of the children?  That's where I think X-Trace's way of recording casuality via event IDs comes very handy. There was a paper on combining span and event model: http://cs.brown.edu/~rfonseca/pubs/leavitt.pdf